{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cb65cd65-cb3b-4ded-8abf-a8dca35c3a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import all python libraries for numpy array calculation, biopython, pandas dataframe,hierarchical clustering, generation dendrogram and for visualization.\n",
    "import os\n",
    "import random as rd\n",
    "import warnings\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sbn\n",
    "import matplotlib as mpl\n",
    "import matplotlib as mp\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from matplotlib import colormaps\n",
    "from matplotlib.figure import Figure\n",
    "from matplotlib.colors import Normalize\n",
    "from dtaidistance.subsequence.dtw import subsequence_alignment\n",
    "from dtaidistance import dtw\n",
    "from dtaidistance import ed\n",
    "from dtaidistance import dtw_visualisation as dtwvis\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from scipy.spatial.distance import euclidean\n",
    "from scipy.cluster.hierarchy import average, dendrogram, linkage\n",
    "import datetime\n",
    "warnings.filterwarnings('ignore')\n",
    "pd.set_option('display.max_rows', 200) \n",
    "\n",
    "\n",
    "###########################User Input########################################\n",
    "DistMeasure='DTW' #Manhattan, DTW or Euclidean\n",
    "window=20\n",
    "offset=2\n",
    "UseThreshold=False\n",
    "TopMatchesNum=20\n",
    "ThresholdLow= 1.55\n",
    "ThresholdHigh=1.75 #1.85 works well for DTW with 25 window and 5 offset,3 works well for Euclidean\n",
    "InputPath=\"C:/Users/realt/Downloads/12 Known Prions.fasta\"\n",
    "OutputDirectory=\"C:/Users/realt/Downloads/PrionTestRuns/\"\n",
    "\n",
    "#############################################################################\n",
    "\n",
    "#The values in the NoteValue array is generated based on physico chemical characteristics - MW and hydrophobicity of amino acids.\n",
    "NoteValue = [0.3064,15.6497,60.7668,52.1362,76.6645,97.0000,-15.8315,-6.8806,5.5470,-2.3381,-23.9905,40.4569,60.1592,27.1418,42.9375,-22.8297,-11.7423,-17.1576,-33.9918,7.0585]\n",
    "\n",
    "#columns array contains amino acids as allowed characters.  \n",
    "columns=['G','A','L','M','F','W','K','Q','E','S','P','V','I','C','Y','H','R','N','D','T']\n",
    "strAllowedChars = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\",\"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "\n",
    "#Generate a dictionary of amino acids as allowed characters and corresponding note value obtained based on physicochemical properties.\n",
    "myDict = dict(zip(columns,NoteValue))\n",
    "\n",
    "#Initialize vectors to store obtained peak values - single peaks and three peaks.\n",
    "ThreePeakOutVectorAll=[]\n",
    "SinglePeakOutVectorAll=[]\n",
    "ThreePeakOutVectorAllP=[]\n",
    "SinglePeakOutVectorAllP=[]\n",
    "\n",
    "#Read and parse test fasta file containing prototypical and prion sequences using biopython module.\n",
    "records = list(SeqIO.parse(InputPath, \"fasta\"))\n",
    "\n",
    "#nameList is an empty list initialized to contain fasta header name of each protein sequence.\n",
    "nameList=[]\n",
    "\n",
    "#waveformsdf is initialized as empty dataframe to store cost matrix of dta distance value between each protein sequence.\n",
    "waveformsdf = pd.DataFrame()\n",
    "\n",
    "PRDRec = pd.DataFrame()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6b0c6061",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions#######################################################\n",
    "#Get the n lowest distance records\n",
    "def lowest(a, n): return np.partition(a, n-1)[:n]\n",
    "\n",
    "def WindowingFunc(seq,window=window, offset=offset):\n",
    "    Searcher=[]\n",
    "    startstopslst=[]\n",
    "    for i in range(len(seq)):\n",
    "        starter=(i*offset)-offset\n",
    "        ender=starter+window\n",
    "        start_stops=str(starter)+'_'+str(ender)\n",
    "        currentsearch=seq[starter:ender]\n",
    "        if len(currentsearch)==window:\n",
    "            Searcher.append(currentsearch)\n",
    "            startstopslst.append(start_stops)\n",
    "    return Searcher,startstopslst\n",
    "\n",
    "#Calculate Manhattan Distance\n",
    "def ManhattanDist(vec1,vec2):\n",
    "    distsum=np.sum(np.abs(vec1,vec2))\n",
    "    return distsum\n",
    "\n",
    "# Get top 10 records by distance\n",
    "def TopMatches(df,TopMatches=TopMatchesNum):\n",
    "    return df.nsmallest(TopMatches, 'Distance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7de7cb04-f1c1-42e0-874c-760cc2dcc431",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Iterate through each protein fasta sequence of test multi fasta file and parse each of them to get fasta header name and fasta sequence.\n",
    "VectorList=[]\n",
    "nameList=[]\n",
    "for i in range(0,len(records)):\n",
    "    #extract fasta header of each protein sequence\n",
    "\n",
    "    ##changing the long description to short id.\n",
    "    names = records[i].id\n",
    "    #Append each fasta header name to the list - nameList.\n",
    "    nameList.append(names)\n",
    "#     print (names)\n",
    "    \n",
    "    #Get the protein sequence as string value into currentseq variable. \n",
    "    currentseq = records[i].seq\n",
    "    \n",
    "    #Count the length of the sequence and store into lengther variable. \n",
    "    lengther=len(currentseq)\n",
    "    \n",
    "\n",
    "    ##This block of code looks for characters that are not amino acids and if any are found replaces them with a 'G'\n",
    "    #Convert all of the charaters to upper before converting it to a list (JD)\n",
    "    currentseq=currentseq.upper()\n",
    "\n",
    "    #Convert the current sequence to a list (JD)\n",
    "    currentseq=list(currentseq)\n",
    "\n",
    "    #Loop through all of the characters and if any are not allowed amino acids, replace them with a 'G'(AK)\n",
    "    for j in range(lengther):\n",
    "            strAllowedChars = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\",\"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "\n",
    "            if currentseq[j] not in strAllowedChars: \n",
    "                currentseq[j] = currentseq[j].replace(currentseq[j],'G')\n",
    "    #Convert the list back to a string (JD)\n",
    "    currentseq=''.join(currentseq)\n",
    "\n",
    "    ##   \n",
    "    \n",
    "    ##This block of code creates a vector containing only the notevalues representing the amino acid sequence (JD).\n",
    "    NoteVec=[]\n",
    "    ender=currentseq[-1]\n",
    "    currentseq=\"G\"+currentseq+ender\n",
    "    lengther=len(currentseq)\n",
    "#     print (currentseq)\n",
    "\n",
    "    for k in range(lengther):\n",
    "        if k==lengther:\n",
    "            NoteVec.append(myDict[currentseq[k-1]])\n",
    "        else:\n",
    "            NoteVec.append(myDict[currentseq[k]])\n",
    "\n",
    "    ##   \n",
    "    \n",
    "    ##This block of code converts a vector of NoteValues to a three-peak vector (JD)\n",
    "    #Get the length of the NoteVector.\n",
    "    lengther2=(len(NoteVec))-1\n",
    "    \n",
    "    #Empty lists are initialized to store the notevalue corresponding to each amino acid.\n",
    "    ThreePeakOutVector=[]\n",
    "    SinglePeakOutVector=[]\n",
    "\n",
    "    #This block of code calculate PreOutvalue,Outvalue,PostOutvalue as three peak values for each amino acid position in protein fasta sequence.\n",
    "    for l in range(lengther2):\n",
    "        \n",
    "        # This block of code is to claculate peak values for the first amino acid of the sequence.\n",
    "        if l==0:\n",
    "            PreOutvalue=NoteVec[l]/2\n",
    "            Outvalue=((NoteVec[l]*2)+((NoteVec[l]+NoteVec[l+1])/2))/2\n",
    "            PostOutvalue=(NoteVec[l]+((NoteVec[l]+NoteVec[l+1])/2))/2\n",
    "            \n",
    "            #Append peak value into the list - ThreePeakOutVector. \n",
    "            ThreePeakOutVector.append(round(PreOutvalue,4))\n",
    "            ThreePeakOutVector.append(round(Outvalue,4))\n",
    "            ThreePeakOutVector.append(round(PostOutvalue,4))\n",
    "            \n",
    "        # This block of code is to claculate peak values for  amino acid of the sequence.\n",
    "        elif i==lengther:\n",
    "            PreOutvalue=(NoteVec[l]+((NoteVec[l]+NoteVec[l-1])/2))/2\n",
    "            Outvalue=((NoteVec[l]*2)+((NoteVec[l]+NoteVec[l-1])/2))/2\n",
    "            PostOutvalue=NoteVec[i]\n",
    "            \n",
    "            #Append peak value into the list - ThreePeakOutVector.\n",
    "            ThreePeakOutVector.append(round(PreOutvalue,4))\n",
    "            ThreePeakOutVector.append(round(Outvalue,4))\n",
    "            ThreePeakOutVector.append(round(PostOutvalue,4))\n",
    "        \n",
    "        # This block of code is to claculate peak values for the rest of the amino acids of the sequence.\n",
    "        else:\n",
    "            PreOutvalue=(NoteVec[l]+((NoteVec[l]+NoteVec[l-1])/2))/2\n",
    "            Outvalue=((NoteVec[l])+((NoteVec[l]+NoteVec[l+1])/2)+((NoteVec[l]+NoteVec[l-1])/2))/2  \n",
    "            PostOutvalue=(NoteVec[l]+((NoteVec[l]+NoteVec[l+1])/2))/2\n",
    "            \n",
    "            #Append peak value into the list - ThreePeakOutVector.\n",
    "            ThreePeakOutVector.append(round(PreOutvalue,4))\n",
    "            ThreePeakOutVector.append(round(Outvalue,4))\n",
    "            ThreePeakOutVector.append(round(PostOutvalue,4)) \n",
    "            \n",
    "    #This extracts the single peak vector from the three peak vector (JD)        \n",
    "    SinglePeakOutVector=ThreePeakOutVector[1::3]\n",
    "    ThreePeakOutVectorAll.append(np.array(ThreePeakOutVector))\n",
    "    SinglePeakOutVectorAll=np.array(SinglePeakOutVector)\n",
    "    \n",
    "    \n",
    "    VectorList.append(SinglePeakOutVectorAll)\n",
    "    \n",
    "RawVecsDF=pd.DataFrame()\n",
    "RawVecsDF['Name']=nameList\n",
    "RawVecsDF['SPVEC']=VectorList\n",
    "\n",
    "# display(RawVecsDF)\n",
    "\n",
    "TempLengthslst=[]\n",
    "for i in range(len(RawVecsDF)):\n",
    "    TempLengthslst.append(len(RawVecsDF.iloc[i]['SPVEC']))\n",
    "    \n",
    "TempNamezlst=list(RawVecsDF.Name)\n",
    "\n",
    "TempLengthsDict=dict(map(lambda i,j : (i,j) , TempNamezlst,TempLengthslst))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7abd7022",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "windowslst=[]\n",
    "vecwindowslst=[]\n",
    "for i in range(len(RawVecsDF)):\n",
    "    windowFullReturn=WindowingFunc(RawVecsDF.iloc[i]['SPVEC'])\n",
    "    windows=windowFullReturn[1]\n",
    "    vecwindows=windowFullReturn[0]\n",
    "    windowslst.append(windows)\n",
    "    vecwindowslst.append(vecwindows)\n",
    "RawVecsDF['Range']=windowslst\n",
    "RawVecsDF['WindowVecs']=vecwindowslst\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bd402385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TempName</th>\n",
       "      <th>TempRange</th>\n",
       "      <th>QueryName</th>\n",
       "      <th>QueryRange</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>5_30</td>\n",
       "      <td>3.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>10_35</td>\n",
       "      <td>3.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>15_40</td>\n",
       "      <td>2.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>0_25</td>\n",
       "      <td>SUP35_P05453</td>\n",
       "      <td>20_45</td>\n",
       "      <td>3.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760924</th>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>205_230</td>\n",
       "      <td>8.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760925</th>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>210_235</td>\n",
       "      <td>4.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760926</th>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>215_240</td>\n",
       "      <td>9.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760927</th>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>220_245</td>\n",
       "      <td>5.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760928</th>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>PRNP_P04156</td>\n",
       "      <td>225_250</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1760929 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TempName TempRange     QueryName QueryRange  Distance\n",
       "0        SUP35_P05453      0_25  SUP35_P05453       0_25      0.00\n",
       "1        SUP35_P05453      0_25  SUP35_P05453       5_30      3.56\n",
       "2        SUP35_P05453      0_25  SUP35_P05453      10_35      3.44\n",
       "3        SUP35_P05453      0_25  SUP35_P05453      15_40      2.55\n",
       "4        SUP35_P05453      0_25  SUP35_P05453      20_45      3.41\n",
       "...               ...       ...           ...        ...       ...\n",
       "1760924   PRNP_P04156   225_250   PRNP_P04156    205_230      8.51\n",
       "1760925   PRNP_P04156   225_250   PRNP_P04156    210_235      4.44\n",
       "1760926   PRNP_P04156   225_250   PRNP_P04156    215_240      9.53\n",
       "1760927   PRNP_P04156   225_250   PRNP_P04156    220_245      5.09\n",
       "1760928   PRNP_P04156   225_250   PRNP_P04156    225_250      0.00\n",
       "\n",
       "[1760929 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "temporarydistlist=[]\n",
    "tempnamelist=[]\n",
    "temprangelist=[]\n",
    "temporarydistlist=[]\n",
    "querynamelist=[]\n",
    "queryrangelist=[]\n",
    "querydistlist=[]\n",
    "\n",
    "for i in range(len(RawVecsDF)):\n",
    "    tempvecnames=RawVecsDF.iloc[i]['Name']\n",
    "    tempvecwindows=RawVecsDF.iloc[i]['WindowVecs']\n",
    "#     print(len(tempvecwindows))\n",
    "    tempvecrange=RawVecsDF.iloc[i]['Range']\n",
    "#     print(len(tempvecrange))\n",
    "    for j in range(len(RawVecsDF)):\n",
    "        queryvecnames=RawVecsDF.iloc[j]['Name']\n",
    "        queryvecwindows=RawVecsDF.iloc[j]['WindowVecs']\n",
    "#         print(len(queryvecwindows))\n",
    "        queryvecrange=RawVecsDF.iloc[j]['Range']\n",
    "#         print(len(queryvecrange))\n",
    "        \n",
    "        \n",
    "        for k in range(len(tempvecwindows)):\n",
    "            tempmeasseqvec=tempvecwindows[k]\n",
    "            tempmeasseqrange=tempvecrange[k]\n",
    "            for l in range(len(queryvecwindows)):\n",
    "                querymeasseqvec=queryvecwindows[l]\n",
    "                querymeasseqrange=queryvecrange[l]\n",
    "                if DistMeasure=='DTW':\n",
    "                    distance=dtw.distance_fast(tempmeasseqvec,querymeasseqvec)\n",
    "                    distance=round(distance/window,2)\n",
    "                elif DistMeasure=='Euclidean':\n",
    "                    distance=ed.distance_fast(tempmeasseqvec,querymeasseqvec)\n",
    "                    distance=round(distance/window,2)\n",
    "                elif DistMeasure=='Manhattan':\n",
    "                    distance=ManhattanDist(tempmeasseqvec,querymeasseqvec)\n",
    "                    distance=round(distance/window,2)\n",
    "                \n",
    "                \n",
    "                tempnamelist.append(tempvecnames)\n",
    "                temprangelist.append(tempmeasseqrange)\n",
    "                \n",
    "                querynamelist.append(queryvecnames)\n",
    "                queryrangelist.append(querymeasseqrange)\n",
    "                temporarydistlist.append(distance)\n",
    "\n",
    "MatchedFragsDF=pd.DataFrame()\n",
    "MatchedFragsDF['TempName']=tempnamelist\n",
    "MatchedFragsDF['TempRange']=temprangelist\n",
    "MatchedFragsDF['QueryName']=querynamelist\n",
    "MatchedFragsDF['QueryRange']=queryrangelist\n",
    "MatchedFragsDF['Distance']=temporarydistlist\n",
    "\n",
    "display(MatchedFragsDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a6725450",
   "metadata": {},
   "outputs": [],
   "source": [
    "if UseThreshold is True:\n",
    "    ThreshFilteredDistDF = MatchedFragsDF[(MatchedFragsDF['Distance'] >= ThresholdLow) & (MatchedFragsDF['Distance'] <= ThresholdHigh)]\n",
    "    ThreshFilteredDistDF=ThreshFilteredDistDF[ThreshFilteredDistDF['TempName']!=ThreshFilteredDistDF['QueryName']]\n",
    "#     display(ThreshFilteredDistDF)\n",
    "elif UseThreshold is False:\n",
    "    ThreshFilteredDistDFpre=MatchedFragsDF[MatchedFragsDF['TempName']!=MatchedFragsDF['QueryName']]\n",
    "    ThreshFilteredDistDF = ThreshFilteredDistDFpre.groupby('TempName', group_keys=False).apply(TopMatches)\n",
    "#     display(ThreshFilteredDistDF)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4be0d1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import Normalize\n",
    "\n",
    "# Get a list of the unique templates in ThreshFilteredDistDF\n",
    "UniqueTemps = np.unique(list(ThreshFilteredDistDF.TempName))\n",
    "# print(UniqueTemps)\n",
    "\n",
    "#Get a timestamp to use in the filenames\n",
    "TimeStamp = datetime.datetime.now().strftime(\"%Y%m%d%H%M%S\")\n",
    "\n",
    "#Make the output directory if it does not exist\n",
    "OutDirectoryFinal=OutputDirectory+str(TimeStamp)+\"/\"\n",
    "if not os.path.exists(OutDirectoryFinal):\n",
    "    os.makedirs(OutDirectoryFinal)\n",
    "    \n",
    "#Get a runtype stamp to use in the filename\n",
    "if UseThreshold is True:\n",
    "    RunType=str(DistMeasure)+\"_Threshold_\"+str(ThresholdLow)+\"_\"+str(ThresholdHigh)\n",
    "elif UseThreshold is False:\n",
    "    RunType=str(DistMeasure)+\"_Top_\"+str(TopMatchesNum)\n",
    "\n",
    "for temp in UniqueTemps:\n",
    "    UniqueTempDF = ThreshFilteredDistDF[ThreshFilteredDistDF['TempName'] == temp]\n",
    "    TempName = temp\n",
    "    XMax = int(TempLengthsDict[temp])\n",
    "    \n",
    "    # Plot settings\n",
    "    figsize_width = 6  # Fixed width\n",
    "    barheight = 0.2  # Consistent height for bars\n",
    "    figsize_height = (len(UniqueTempDF) + 1) * barheight  # Adjust height based on number of bars\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(figsize_width, figsize_height))\n",
    "    \n",
    "    # Normalize distance for color mapping\n",
    "    norm = Normalize(vmin=min(ThreshFilteredDistDF.Distance), vmax=max(ThreshFilteredDistDF.Distance))\n",
    "    cmap = cm.get_cmap('plasma')\n",
    "    sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    \n",
    "    # Add the template bar at the top\n",
    "    ax.barh(y=barheight * len(UniqueTempDF), width=XMax, height=barheight, color='Black')\n",
    "    \n",
    "    ticklabels = [TempName]\n",
    "    ticklocations = [barheight * len(UniqueTempDF)]\n",
    "    \n",
    "    # Add the fragments\n",
    "    for j in range(len(UniqueTempDF)):\n",
    "        distance = UniqueTempDF.iloc[j]['Distance']\n",
    "        start = int(UniqueTempDF.iloc[j]['TempRange'].split(\"_\")[0])\n",
    "        width = int(UniqueTempDF.iloc[j]['TempRange'].split(\"_\")[1]) - start  # Actual width of fragment\n",
    "        yloc = barheight * j#(len(UniqueTempDF) - j - 1)\n",
    "        color = cmap(norm(distance))\n",
    "        \n",
    "        ax.barh(yloc, width, left=start, height=barheight, color=color)\n",
    "        \n",
    "        fullname = f\"{UniqueTempDF.iloc[j]['QueryName']}_{UniqueTempDF.iloc[j]['QueryRange']}\"\n",
    "        ticklabels.append(fullname)\n",
    "        ticklocations.append(yloc)\n",
    "        \n",
    "    ax.margins(x=0, y=0)\n",
    "    ax.set_xlim(0, XMax)\n",
    "    ax.set_yticks(ticklocations)\n",
    "    ax.set_yticklabels(ticklabels, fontsize=6)\n",
    "    \n",
    "    # Remove padding\n",
    "    plt.subplots_adjust(left=0, right=1, top=1, bottom=0.3)  # Adjusted bottom to leave space for colorbar\n",
    "        \n",
    "    plt.savefig(OutDirectoryFinal+TempName+\"_\"+RunType+\"_win_\"+str(window)+\"_off_\"+str(offset)+\".png\", bbox_inches='tight', dpi=300)\n",
    "    plt.close()\n",
    "\n",
    "# Create a standalone color bar that you can add to an aggregate figure later in Publisher, Illustrator, etc.\n",
    "fig, ax = plt.subplots(figsize=(6, .125))  #Size as needed\n",
    "cbar_ax = fig.add_axes([0.3, 0.3, 0.4, 0.5])  #Position/size\n",
    "fig.colorbar(sm, cax=cbar_ax, orientation=\"horizontal\", label='Distance')\n",
    "ax.set_axis_off()\n",
    "plt.savefig(OutDirectoryFinal+\"ColorBar\"+\"_\"+RunType+\"_win_\"+str(window)+\"_off_\"+str(offset)+\".png\", bbox_inches='tight', dpi=300)\n",
    "plt.close()\n",
    "\n",
    "MatchedFragsDF.to_pickle(OutDirectoryFinal+\"AllMatchesDF\"+\"_\"+DistMeasure+\"_win_\"+str(window)+\"_off_\"+str(offset)+\".pkl\")\n",
    "ThreshFilteredDistDF.to_pickle(OutDirectoryFinal+\"PlottedResultsDF\"+\"_\"+DistMeasure+\"_win_\"+str(window)+\"_off_\"+str(offset)+\".pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81dd286-949c-4f78-a4f2-25ebdd39bdc4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
